token-lending/program/src/state/reserve.rs

```diff
@@ -565,50 +631,67 @@ impl ReserveFees {
     /// Calculate the owner and host fees on borrow
     pub fn calculate_borrow_fees(
         &self,
-        collateral_amount: u64,
+        borrow_amount: Decimal,
+        fee_calculation: FeeCalculation,
     ) -> Result<(u64, u64), ProgramError> {
         let borrow_fee_rate = Rate::from_scaled_val(self.borrow_fee_wad);
         let host_fee_rate = Rate::from_percent(self.host_fee_percentage);
-        if borrow_fee_rate > Rate::zero() && collateral_amount > 0 {
+        if borrow_fee_rate > Rate::zero() && borrow_amount > Decimal::zero() {
             let need_to_assess_host_fee = host_fee_rate > Rate::zero();
             let minimum_fee = if need_to_assess_host_fee {
                 2 // 1 token to owner, 1 to host
             } else {
                 1 // 1 token to owner, nothing else
             };
 
-            let borrow_fee = borrow_fee_rate
-                .try_mul(collateral_amount)?
-                .try_round_u64()?
-                .max(minimum_fee);
+            let borrow_fee_amount = match fee_calculation {
+                FeeCalculation::Exclusive => borrow_amount.try_mul(borrow_fee_rate)?,
+                FeeCalculation::Inclusive => {
+                    let borrow_fee_rate =
+                        borrow_fee_rate.try_div(borrow_fee_rate.try_add(Rate::one())?)?;
+                    borrow_amount.try_mul(borrow_fee_rate)?
```

Comment:
Looks correct, but do you mind writing some algebra in comments to explain this how you got to this?

Reply:
For sure, those formulas are here: https://github.com/solana-labs/solana-program-library/blob/d371cceda5321b46e7c6cd468d03dd71e4f1930b/token-lending/program/src/state/reserve.rs#L675-L681

It seemed better to keep them in one place but I can inline them if you still want?

Reply:
I did see these, I was just hoping for something inline that explained that we are calculating a fee such that when it's applied to the received amount, is equal to the specified borrow amount.

Reply:
Cool, can do.